scripts/util/charmhub.py

```diff
@@ -0,0 +1,198 @@
+    k8s and k8s-worker charms
+    """
+    
+    def __init__(self):
+        self.data: defaultdict[str, RevisionMatrix] = defaultdict(None)
+
+    def set(self, charm, revision_matrix):
+        self.data[charm] = revision_matrix
+
+    def is_testable(self):
+        if not len(self.data) or any(matrix is None for matrix in self.data.values()):
+            return False
+        
+        # All the matrices in a bundle must have the same span of arch and bases 
+        # and have a revision values for each (arch, base) so that they can be 
+        # tested alongside each other.
+        item: RevisionMatrix = random.choice(list(self.data.values())) # nosec
+        
+        bases = item.get_bases()
+        archs = item.get_archs()
+
+        for revision_matrix in self.data.values():
+            if revision_matrix.get_bases() != bases or revision_matrix.get_archs() != archs:
+                return False
+            
+            for base in bases:
+                for arch in archs:
+                    if item.get(arch, base) and not revision_matrix.get(arch, base):
+                            return False
+                    
+        return True
+
+    def get_bases(self):
+        try:
+            item: RevisionMatrix = random.choice(list(self.data.values())) # nosec
+            return item.get_bases()
+        except StopIteration:
+            return set()
+        
+    def get_archs(self):
+        try:
+            item: RevisionMatrix = random.choice(list(self.data.values())) # nosec
+            return item.get_archs()
+        except StopIteration:
+            return set()
+    
+    def get_revisions(self, arch, base):
+        revisions = {}
+
+        for charm in self.data.keys():
+            revisions[f"{charm.replace("-", "_")}_revision"] = self.data[charm].get(arch, base)
+
+        return revisions
+
+    def get_version(self, arch, base):
+        
+        charms = sorted(self.data.keys())
+        if not charms:
+            return None
+
+        version = "k8s-operator"
+        for charm in charms:
+            revision_matrix = self.data[charm]
+            if not revision_matrix:
+                return None
+                
+            revision = revision_matrix.get(arch, base)
+            if not revision:
+                return None
+
+            version += f"-{charm}-{revision}"
+
+        return version
+
+class RevisionMatrix:
+    """
+    For each tuple of (name, channel, arch, base) there is a unique charm artifact
+    in Charmhub. RevisionMatrix is a matrix of (arch, base) revisions, if any, for
+    a specific (name, channel) tuple.
+    Rows of the matrix correspond to different architectures.
+    Columns of the matrix correspond to different bases.
+    """
+
+    def __init__(self):
+        self.data: defaultdict[str, dict[str, str]] = defaultdict(dict)
+
+    def set(self, arch, base, revision):
+        self.data[arch][base] = revision
```

Comment:
why not just do this, the nested dict makes alot of things confusing

```suggestion
    def set(self, arch, base, revision):
        self.data[(arch, base)] = revision
```
```

Reply:
Done, thanks a lot